src/combinatorics/young_diagram.lean

```diff
@@ -0,0 +1,101 @@
+to say that `(i, j)` (in matrix coordinates) is in the Young diagram `μ`.
+
+## Main definitions:
+
+- `young_diagram` : Young diagrams
+- `young_diagram.card` : the number of cells in a Young diagram (its *cardinality*)
+- `young_diagram.distrib_lattice` : a distributive lattice instance for Young diagrams ordered by containment, with `(⊥ : young_diagram)` the empty diagram.
+
+## Notation:
+
+In "English notation", a Young diagram is drawn so that (i1, j1) ≤ (i2, j2)
+means (i1, j1) is weakly up-and-left (northwest) of (i2, j2). This terminology is used
+below, e.g. in `young_diagram.nw_of`.
+
+## Tags
+
+Young diagram
+
+## References
+
+<https://en.wikipedia.org/wiki/Young_tableau>
+
+-/
+
+section young_diagram
+
+/-- A Young diagram is a lower set in ℕ × ℕ -/
+@[ext] structure young_diagram :=
+(cells : finset (ℕ × ℕ))
+(is_lower_set : is_lower_set (cells : set (ℕ × ℕ)))
+
+instance young_diagram.set_like : set_like young_diagram (ℕ × ℕ) :=
+{ coe            := λ μ, (μ.cells : set (ℕ × ℕ)),
+  coe_injective' := λ μ ν h, by { rwa [young_diagram.ext_iff, ← finset.coe_inj] } }
+
+@[simp] lemma young_diagram.mem_cells {μ : young_diagram} (c : ℕ × ℕ) :
+  c ∈ μ.cells ↔ c ∈ μ := iff.rfl
+
+/-- In "English notation", a Young diagram is drawn so that (i1, j1) ≤ (i2, j2)
+    means (i1, j1) is weakly up-and-left (northwest) of (i2, j2). -/
+lemma young_diagram.nw_of (μ : young_diagram) {i1 i2 j1 j2 : ℕ}
+  (hi : i1 ≤ i2) (hj : j1 ≤ j2) (hcell : (i2, j2) ∈ μ) : (i1, j1) ∈ μ :=
+μ.is_lower_set (prod.mk_le_mk.mpr ⟨hi, hj⟩) hcell
+
+section distrib_lattice
+
+instance young_diagram.has_le : has_le young_diagram :=
+{ le := λ μ ν, (μ : set (ℕ × ℕ)) ⊆ ν }
+
+instance young_diagram.has_sup : has_sup young_diagram :=
+{ sup := λ μ ν, { cells        := μ.cells ∪ ν.cells,
+                  is_lower_set := by { rw finset.coe_union,
+                                       exact is_lower_set.union μ.is_lower_set ν.is_lower_set } } }
+
+instance young_diagram.has_inf : has_inf young_diagram :=
+{ inf := λ μ ν, { cells        := μ.cells ∩ ν.cells,
+                  is_lower_set := by { rw finset.coe_inter,
+                                       exact is_lower_set.inter μ.is_lower_set ν.is_lower_set } } }
+
+/-- The empty Young diagram is (⊥ : young_diagram). -/
+instance young_diagram.has_bot : has_bot young_diagram :=
+{ bot := { cells := finset.empty, is_lower_set := λ _ _ _ h, h } }
+
+instance young_diagram.distrib_lattice : distrib_lattice young_diagram :=
+function.injective.distrib_lattice
+  (λ (μ : young_diagram), μ.cells)
+  (λ μ ν h, by rwa young_diagram.ext_iff)
+  (λ _ _, rfl) (λ _ _, rfl)
+
+instance : inhabited young_diagram := ⟨⊥⟩
+
+end distrib_lattice
+
+/-- Cardinality of a Young diagram -/
+def young_diagram.card (μ : young_diagram) : ℕ := μ.cells.card
+lemma young_diagram.card_def (μ : young_diagram) : μ.card = μ.cells.card := rfl
+
+@[simp] lemma young_diagram.card_bot : (⊥ : young_diagram).card = 0 := rfl
```

Comment:
I was worrying about this, too, and I'm not sure what the right thing to do is. Maybe it won't be so much duplication since we're not doing unrestricted set operations with Young diagrams? For example, we don't need anything about the cardinality of the sup of two disjoint Young diagrams since that's impossible.

Maybe leave it for now and come back to it if we start seeing a recapitulation of the finset API?

Reply:
Oh I see. Is there a way to have some kind of coercion? The main thing is it would be nice to have dot notation and be able to just say `μ.card` rather than `μ.cells.card` or `finset.card μ`. Actually the latter doesn't work right now since currently there is no coercion from `young_diagram` to `finset` (which, cardinality aside, seems like it may be annoying down the road).

Reply:
Ideally I would like to suppress all reference to `μ.cells`. This is how I did it in the library I've built so far.

Reply:
Actually, this reminds me of @YaelDillies's suggestion about `lower_finset`, except that this seems like it could apply to anything "`finset_like`".

Reply:
I can add `@[reducible]` for now, though it looks like it won't prove `(⊥ : young_diagram).card = 0` by `simp`. Maybe because the `has_bot.bot` definition is also not reducible?

Reply:
> I can add `@[reducible]` for now, though it looks like it won't prove `(⊥ : young_diagram).card = 0` by `simp`. Maybe because the `has_bot.bot` definition is also not reducible?

It does now prove it by `simp` now that you added the lemmas about `bot.cells`!